stablehlo/integrations/python/stablehlo/savedmodel/stablehlo_to_tf_saved_model.py

```diff
@@ -0,0 +1,246 @@
+      if str(loc.type_) == str(VariableType.PARAMETER):
+        call_args.append(self.stablehlo_program.state_dict[loc.name])
+      else:
+        call_args.append(args[loc.position])
+    return call_args
+
+  def _wrap_as_tf_func(self):
+    def inner(*args):
+      Touts = [
+          self.stablehlo_type_to_tf_type[sig.dtype]
+          for sig in self.stablehlo_program.output_signature
+      ]
+      Souts = [
+          self._get_shape_with_dynamic(sig)
+          for sig in self.stablehlo_program.output_signature
+      ]
+      call_args = self._extract_call_parameters(args)
+      m = tfxla.call_module(
+          tuple(call_args),
+          version=5,
+          Tout=Touts,  # dtype information
+          Sout=Souts,  # Shape information
+          function_list=[],
+          module=self.stablehlo_program.bytecode,
+      )
+      return m
+
+    return inner
+
+  def _make_tf_function(self):
+    return self._wrap_as_tf_func()
+
+  def _make_input_signatures(self) -> List[tf.TensorSpec]:
+    input_pos_to_spec = {
+        loc.position: spec
+        for loc, spec in itertools.chain(
+            zip(
+                self.stablehlo_program.input_locations,
+                self.stablehlo_program.input_signature,
+            ),
+            [],
+        )
+        if str(loc.type_) == str(VariableType.INPUT_ARG)
+    }
+    for i in range(len(input_pos_to_spec)):
+      spec = input_pos_to_spec[i]
+      shape = self._get_shape_with_dynamic(spec)
+      yield tf.TensorSpec(
+          shape=shape,
+          dtype=getattr(
+              tf,
+              self.stablehlo_type_to_tf_type[spec.dtype]
+              if spec.dtype in self.stablehlo_type_to_tf_type
+              else spec.dtype,
+          ),
+          name=f'args_{i}',
+      )
+
+  def to_tf_saved_model(
+      self,
+      path: os.PathLike,
+      serving_key: str = tf.saved_model.DEFAULT_SERVING_SIGNATURE_DEF_KEY,
+      function_alias: str = '',
+  ) -> None:
+    tfm = tf.Module()
+
+    self.stablehlo_program.state_dict = {
+        k: tf.Variable(v, trainable=False, name=k)
+        for k, v in self.stablehlo_program.state_dict.items()
+    }
+
+    input_signatures = list(self._make_input_signatures())
+
+    tfm.f = tf.function(
+        self._make_tf_function(), input_signature=input_signatures
+    )
+    tfm._variables = list(self.stablehlo_program.state_dict.values())
+    signatures = {serving_key: tfm.f.get_concrete_function(*input_signatures)}
+    save_options = tf.saved_model.SaveOptions(
+        function_aliases={
+            function_alias: tfm.f,
+        }
+    )
+    tf.saved_model.save(
+        tfm,
+        path,
+        signatures=signatures,
+        options=save_options,
+    )
+
+
+# Top level API for stablehlo to tf saved model
+
+
+def stablehlo_to_tf_saved_model(
+    module: ir.Module,
+    saved_model_dir: os.PathLike,
+    input_locations: list = [],
+    state_dict: dict = {},
+):
+  target = stablehlo.get_current_version()
```

Comment:
I think TF has some forward compat.. 1month at least. We could probably make this an input argument that defaults to current